enhancements/compliance-operator.md

```diff
@@ -0,0 +1,222 @@
+and [NIST-certified](https://csrc.nist.gov/projects/security-content-automation-protocol-validation-pr/validated-products-and-modules/142-red-hat-scap-1-2-product-validation-record)
+project that administrators are aware of and trust.
+
+For the first iteration, the operator would only provide a report with
+the findings and proposed remediations which the cluster administrator
+can then apply on their own. Remediation would be developed in one
+of the future versions.
+
+## Motivation
+
+The default installation of OKD is not compliant (and is not expected
+to be, as some requirements have impact on performance or flexibility of use
+of the cluster) with different standards users in regulated environments
+must adhere to.
+
+Enabling users to be compliant with standards their respective industry
+requires would enable OKD to be used in those environments. This is especially
+true of the US public sector.
+
+### Dependencies
+As said above, the operator would orchestrate OpenSCAP scans on cluster
+nodes. However, there are several areas where OpenSCAP needs to be extended
+to be better usable for scanning an OKD cluster:
+
+* OpenSCAP did not support merging of results from several nodes into a single
+  bundle. Each scan on each node produces a separate report. There is an
+  RFE filed against OpenSCAP and planned by the OpenSCAP team to aggregate
+  results from all the cluster nodes into a single report.
+* The remediations were typically done using a shell script or Ansible. For
+  OpenShift we’d want the remediations done using MachineConfigs or other
+  Custom Resources. However, the remediations are not a hard prerequisite for
+  the first version of the operator, even identifying the gaps in compliance
+  would be useful.
+* For the gaps in compliance we identify, a check must be written. These
+  checks are being compiled as a parallel effort. Having a full set of
+  checks and/or remediations is not a hard prerequisite for the operator
+  itself, though as the content would be delivered separately.
+
+These have all been filed with the OpenSCAP team and acknowledged by them.
+
+### Goals
+
+A cluster administrator is able to use the operator to assess the degree
+of compliance against a given security standard. In particular this would
+mean:
+
+* Provide a way for cluster administrator to start cluster scans
+* Inspect the scan result on the high level (pass/fail/error)
+* Provide a way to collect the detailed results of the scan
+  * The results would list a way to remediate the gaps, typically by the means
+    of deploying certain manifests to OKD (e.g. a Machineconfig), a different
+    way of changing cluster configuration or even a free-form recommendation text.
+
+### Non-Goals
+
+The operator itself does not perform any scans or remediations, only
+wraps the OpenSCAP scanner and provides an OKD-native API to
+run the scans.
+
+## Proposal
+
+An instance of a scan would be represented by a CRD. The CRD would include
```

Comment:
Do all nodes get checked using the same set of rules or it can be customized?

Reply:
It can be customized. There are a couple of ways of doing this (none of them are right or wrong according to the OpenSCAP team). you can create a profile per role (e.g. one profile for masters and one profile for workers), or you can create a general-purpose compliance checks that merely ignores the check if it doesn't apply (e.g. if you need to check for a service that only exists on the master nodes, then you detect that the service isn't there, and ignore the check on the workers).

Reply:
Ah, that makes sense. Seems CRD should mention MachineSet it would operate on then

Reply:
Yeah, probably. Initially I was thinking just nodeSelectors and match against labels, but if a MachineSet makes more sense, then sure, it's an option. (I need to familiarize myself with MachineSets a bit though before I incorporate that into the proposal..)

Reply:
Hi @vrutkovs would you mind helping me expand on the idea? As far as I cans see in my cluster, the MachineSets only reference worker nodes by default, even though [the documentation](https://docs.openshift.com/container-platform/4.1/machine_management/creating-infrastructure-machinesets.html) suggests to create MachineSets for infrastructure components as well. Also, from the points of view of the compliance operator, you'd typically only want to match against labels, typically `node-role` or perhaps `node.openshift.io/os_id`.

So at the moment it is not clear to me how exactly would you suggest to use MachineSets, or what would that give over using something like:
```
nodeSelector:
   node-role.kubernetes.io/worker=
```